[JIT] Resolve string literal type annotations using `Resolver::resolveType`

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47731 [JIT] Resolve string literal type annotations using `Resolver::resolveType`**

**Summary**
This commit modifies `ScriptTypeParser::parseTypeFromExpr` so that
string literal type annotations are resolved using
`Resolver::resolveType`. At present, they are parsed in
`parseBaseTypeName`, which inadvertently allows any key from
`string_to_type_lut` to be used as a string literal type annotation.

**Test Plan**
Existing unit tests (most notably
`TestClassType.test_self_referential_method` which tests the main
feature, self-referential class type annotations, that make use of
string literal type annotations).

**Fixes**
This commit fixes #47570.

Differential Revision: [D24934717](https://our.internmc.facebook.com/intern/diff/D24934717)

Commit:
[JIT] Resolve string literal type annotations using `Resolver::resolveType`

**Summary**
This commit modifies `ScriptTypeParser::parseTypeFromExpr` so that
string literal type annotations are resolved using
`Resolver::resolveType`. At present, they are parsed in
`parseBaseTypeName`, which inadvertently allows any key from
`string_to_type_lut` to be used as a string literal type annotation.

**Test Plan**
Existing unit tests (most notably
`TestClassType.test_self_referential_method` which tests the main
feature, self-referential class type annotations, that make use of
string literal type annotations).

**Fixes**
This commit fixes #47570.

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -190,6 +187,15 @@ TypePtr ScriptTypeParser::parseTypeFromExprImpl(const Expr& expr) const {
     return subscriptToType(*value_name, subscript);
 
+  } else if (expr.kind() == TK_STRINGLITERAL) {
+    auto type_name = StringLiteral(expr).text();
```
nit and optional:

Though `auto` will be deduced correctly, it is better to still include `const` and `&` for better readability.

Commit:
Update on "[JIT] Resolve string literal type annotations using `Resolver::resolveType`"


**Summary**
This commit modifies `ScriptTypeParser::parseTypeFromExpr` so that
string literal type annotations are resolved using
`Resolver::resolveType`. At present, they are parsed in
`parseBaseTypeName`, which inadvertently allows any key from
`string_to_type_lut` to be used as a string literal type annotation.

**Test Plan**
Existing unit tests (most notably
`TestClassType.test_self_referential_method` which tests the main
feature, self-referential class type annotations, that make use of
string literal type annotations).

**Fixes**
This commit fixes #47570.

[ghstack-poisoned]

jeffdaily commented:
ROCm CI failure due to CI script update. Will retest.

jeffdaily commented:
@pytorchbot retest this please

ppwwyyxx commented:
Oops I thought supporting "Device" in https://github.com/pytorch/pytorch/issues/47570 is a feature, not bug :laughing: 
Now the problem is, with this "feature" removed, and because of https://github.com/pytorch/pytorch/issues/47405, we can no longer annotate a method that uses device type in argument. So it actually causes a regression.

SplitInfinity commented:
I fixed #47405 in #47734.

ppwwyyxx commented:
That's great! Didn't notice this fix is also coming
